llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -285,6 +285,13 @@ void collectCompositeElementsInfoRecursive(
   }
 }
 
+Constant *getConstantElement(Constant *C, unsigned i, Type *elementType) {
```

Comment:
Could you please share more insights about what exactly goes wrong here? I'm wondering, what went wrong, because according to [langref](https://llvm.org/docs/LangRef.html#complex-constants), `zeroinitializer` should work fine here:

> Zero initialization
> The string ‘zeroinitializer’ can be used to zero initialize a value to zero of any type, including scalar and aggregate types. This is often used to avoid having to print large zero initializers (e.g. for large arrays) and is always exactly equivalent to using explicit zero initializers.

Unless I'm missing something (like `getNullValue` doesn't generate `zeroinitializer`)

Reply:
Please disregard my comment above, `zeroinitializer` works fine, but we can't just call `getOperand(I)` on zero-initialized constant. Your fix is essentially a continuation of 3023528cd9a2827fd022ef781529884a608ffca3. right?

Reply:
The problem was the following, when we were trying to get `getOperand` in a `zeroinitializer` array: `C->getOperand(i)`, the assert from [User.h](https://github.com/intel/llvm/blob/sycl/llvm/include/llvm/IR/User.h#L170) was triggered: 
```c++
    assert(i < NumUserOperands && "getOperand() out of range!");
```
I believe this is because `zeroinitializer` has no operands.

So the problem occurs only when the specialization constant was not initialized in the point of definition:

```c++
coeff_t get_coefficients() {
  return {{{1.0, 2.0, 3.0}, {1.1, 2.1, 3.1}, {1.2, 2.2, 3.2}}};
}

constexpr specialization_id<coeff_t> coeff_id;

...
    cgh.set_specialization_constant<coeff_id>(get_coefficients());
    cgh.parallel_for<class Convolution>(
    ...
    );
```

When the specialization constant was created as an initialized value, all the things worked as expected

```c++
constexpr coeff_t get_coefficients() {
  return {{{1.0, 2.0, 3.0}, {1.1, 2.1, 3.1}, {1.2, 2.2, 3.2}}};
}
...
constexpr specialization_id<coeff_t> coeff_id{get_coefficients()};
```

Reply:
@AlexeySachkov 

> Your fix is essentially a continuation of 3023528. right?

I believe so.

Reply:
Ok, thanks!